### ISSUE TITLE
Improving sync resilience and logging

### ISSUE THREAD

**Issue by andreievg** [openmsupply/mobile#1027]
## Describe the feature

Need to make sync more resilient with better logging

Bad sync records (form server and from mobile) shouldn't stop sync entirely, but shouldn't just be ignored either

Related issue:  https://github.com/openmsupply/mobile/issues/1025

@Chris-Petty @wlthomson @mark-prins @craigdrown @joshxg @sworup 
Thoughts ?

**Comment by joshxg** [comment]
Another related issue here #210 

With a nice idea, in my opinion here: https://github.com/openmsupply/mobile/issues/210#issue-166993125

Could be extended:

When a sync error occurs, reduce the number of records to one, to find the culprit. Display in mobile, just incase, log on desktop, and send to THE HUB

**Comment by wlthomson** [comment]
Also related issue #987.

Maybe missing something, but wrt. mobile (possibly also desktop, but I don't know much about the mobile->desktop sync API), I'm not sure if it's necessary to reduce the number of records to find culprits for sync errors. AFAIK, records are synced in batches, but still processed and sanity checked individually.

To catch malformed errors implies we can identify them, so I think sync error logging could form part of a refactor of existing sanity checks for incoming records?

**Comment by joshxg** [comment]
So there does seem to be a bit of work here and a bit of thought into how to handle errors happening in sync.

>Bad sync records (form server and from mobile) shouldn't stop sync entirely,

Currently, you can keep spamming sync and roll the dice on which records are coming in hot, such that you're left with few/one of the records that are broken. Technically, we could just make this easier - log the offending record and allow sync to continue. However is that really the behaviour we want?

Do we want a `Transaction` to be edited if `TransactionBatch` records aren't all synced? A `MasterList` to be used which doesn't have all it's `Item` records? Finalize a customer requisition without all it's lines? If no, then really the solution is a bit bigger?

- Ensuring any record and it's related records are valid *before* sending (desktop and mobile)
- Fetching changes such that the 'puller' knows exactly what they're pulling - pull a `transact` and all it's `trans_line` records, validate and then integrate. Most likely responding with why records pulled were rejected

**Comment by wlthomson** [comment]
> So there does seem to be a bit of work here and a bit of thought into how to handle errors happening in sync.
> 
> > Bad sync records (form server and from mobile) shouldn't stop sync entirely,
> 
> Currently, you can keep spamming sync and roll the dice on which records are coming in hot, such that you're left with few/one of the records that are broken. Technically, we could just make this easier - log the offending record and allow sync to continue. However is that really the behaviour we want?
> 
> Do we want a `Transaction` to be edited if `TransactionBatch` records aren't all synced? A `MasterList` to be used which doesn't have all it's `Item` records? Finalize a customer requisition without all it's lines? If no, then really the solution is a bit bigger?
> 
>     * Ensuring any record and it's related records are valid _before_ sending (desktop and mobile)
> 
>     * Fetching changes such that the 'puller' knows exactly what they're pulling - pull a `transact` and all it's `trans_line` records, validate and then integrate. Most likely responding with why records pulled were rejected

See https://github.com/openmsupply/mobile/issues/1088#issuecomment-518416585.

TLDR; my opinion (which I think is the same as yours) is that we need to enforce data integrity of existing records (in mobile and desktop) rather than pushing this responsibility off to when it becomes a problem (i.e. sync errors). 

Of course, still a role for integrity checking in incoming sync, but it shouldn't be relied upon as a defacto data fixer-upper.
